src/hpke.rs

```diff
@@ -0,0 +1,602 @@
+                let serialized_key =
+                    keystore.load_key(key::Secrecy::Secret, None, &req.key_to_seal)?;
+                let mut message = Message::from_slice(&serialized_key.serialize()).unwrap();
+
+                let public_key = load_public_key(&req.public_key, keystore)?;
+
+                let (pk, tag) = seal8(
+                    public_key,
+                    &req.info,
+                    &req.aad,
+                    &mut message,
+                    keystore.rng(),
+                );
+
+                message
+                    .extend_from_slice(&pk.to_bytes())
+                    .map_err(|_| trussed::Error::SignDataTooLarge)?;
+                message
+                    .extend_from_slice(&tag)
+                    .map_err(|_| trussed::Error::SignDataTooLarge)?;
+
+                Ok(HpkeSealKeyReply { data: message }.into())
+            }
+            HpkeRequest::SealKeyToFile(req) => {
+                // TODO: need to check both secret and private keys
+                let serialized_key =
+                    keystore.load_key(key::Secrecy::Secret, None, &req.key_to_seal)?;
+                let mut message = Bytes::<{ MAX_SERIALIZED_KEY_LENGTH + 32 + 16 }>::from_slice(
+                    &serialized_key.serialize(),
+                )
+                .unwrap();
+
+                let public_key = load_public_key(&req.public_key, keystore)?;
+
+                let (pk, tag) = seal8(
+                    public_key,
+                    &req.info,
+                    &req.aad,
+                    &mut message,
+                    keystore.rng(),
+                );
+
+                message
+                    .extend_from_slice(&pk.to_bytes())
+                    .map_err(|_| trussed::Error::SignDataTooLarge)?;
+                message
+                    .extend_from_slice(&tag)
+                    .map_err(|_| trussed::Error::SignDataTooLarge)?;
+                filestore.write(&req.file, req.location, &message)?;
+
+                Ok(HpkeSealKeyToFileReply {}.into())
+            }
+            HpkeRequest::Open(req) => {
+                let enc = load_public_key(&req.enc_key, keystore)?;
+                let secret_key = load_secret_key(&req.key, keystore)?;
+
+                let mut ct = req.ciphertext.clone();
+                open8(
+                    enc,
+                    secret_key,
+                    &req.info,
+                    &req.aad,
+                    &mut ct,
+                    req.tag.into(),
+                )
+                .map_err(|_| trussed::Error::AeadError)?;
+
+                Ok(HpkeOpenReply { plaintext: ct }.into())
+            }
+            HpkeRequest::OpenKey(req) => {
+                let secret_key = load_secret_key(&req.key, keystore)?;
+                let mut ct = req.sealed_key.clone();
+                let (ct, tag) = ct.split_last_chunk_mut().ok_or(trussed::Error::AeadError)?;
```

Comment:
good chances I don't fully understand this - but will this also work if `ct` is very short, i.e., shorter than `TAG_LEN` ?

Reply:
`split_at_mut` does not panic if `ct` is too small. It returns `None`,